### PR TITLE
commit with search feature working

### PR DESCRIPTION
**Pull request by AbelKel**
- Resolved search bar error issue. The search bar should now be working.
- Added a new View Controller (NewsViewController) and  (NewsCell) with API calls to News API.
Image 7-8-22 at 4 52 AM

### CODE REVIEW

File: Capstone/Authentication/LoginViewController.m [hunk 1]
```diff
@@ -15,6 +15,8 @@
 @interface LoginViewController ()<RegistrationViewControllerDelegate>
 @property (weak, nonatomic) IBOutlet UITextField *usernameLogin;
 @property (weak, nonatomic) IBOutlet UITextField *passwordLogin;
+@property (weak, nonatomic) IBOutlet UIView *fbLogin;
+
```
Comment (skhezri) [inline Capstone/Authentication/LoginViewController.m]: nit: extra line

File: Capstone/Authentication/LoginViewController.m [hunk 2]
```diff
@@ -38,33 +39,27 @@ - (IBAction)didTapLogin:(id)sender {
...
         } else {
             NSLog(@"User logged in successfully");
-            
         }
     }];
 }
 
 
 #pragma mark - FBSDKLoginButton
-
```
Comment (skhezri) [inline Capstone/Authentication/LoginViewController.m:59]: great job resolving these extra spaces!!

File: Capstone/Feed/HomeViewController.m [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        for (NSDictionary<NSString *, NSString *> *college in self.colleges) {
+            NSRange collegeName = [college[@"name"] rangeOfString:searchText options:NSCaseInsensitiveSearch];
+            if (collegeName.location != NSNotFound) {
+                [self.filteredColleges addObject:college];
+            }
+        }
+    }
+    [self.tableView reloadData];
+}
+
```
Comment (skhezri) [inline Capstone/Feed/HomeViewController.m:62]: nit: extra line

File: Capstone/News/NewsViewController.m [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration] delegate:nil delegateQueue:[NSOperationQueue mainQueue]];
+    NSURLSessionDataTask *task = [session dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+           if (error != nil) {
+               NSLog(@"%@", [error localizedDescription]);
+           } else {
+               NSDictionary *dataDictionary = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableContainers error:nil];
+               self.collegeNews = [dataDictionary objectForKey:@"articles"];
+               for (NSDictionary<NSString *, NSString *> *college in self.collegeNews) {
+                   [self.college1 addObject:college];
+                   NSLog(@"%@", college);
```
Comment (skhezri) [inline Capstone/News/NewsViewController.m]: nit: remove logging code added for testing purposes

File: Capstone/News/NewsViewController.m [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+               }
+               [self.tableView reloadData];
+            }}];
+    [task resume];
+    NSLog(@"%@123", self.college1);
```
Comment (skhezri) [inline Capstone/News/NewsViewController.m]: nit: remove logging code added for testing purposes

File: Capstone/News/NewsViewController.m [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+//  Created by Abel Kelbessa on 7/7/22.
+//
+
+#import "NewsViewController.h"
+#import "UIImageView+AFNetworking.h"
+#import "NewsCell.h"
+
+@interface NewsViewController ()<UITableViewDelegate,UITableViewDataSource>
+@property(strong, nonatomic) NSMutableArray<NSDictionary *> *collegeNews;
+@property(strong, nonatomic) NSMutableArray *college1;
```
Comment (skhezri) [inline Capstone/News/NewsViewController.m]: let's try and make this variable name more descriptive!

File: Capstone/Feed/HomeViewController.m [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
+    UITableViewCell *cell = sender;
+    NSIndexPath *myIndexPath = [self.tableView indexPathForCell:cell];
+    NSString *nameToPass = self.collegeDictionaryArrays[myIndexPath.row][@"name"];
+    NSString *detailToPass = self.collegeDictionaryArrays[myIndexPath.row][@"shortDescription"];
+    NSString *cityToPass = self.collegeDictionaryArrays[myIndexPath.row][@"city"];
+    NSString *imageToPass = self.collegeDictionaryArrays[myIndexPath.row][@"campusImage"];
+    DetailsViewController *detailVC = [segue destinationViewController];
+    detailVC.detailName = nameToPass;
```
Comment (skhezri) [inline Capstone/Feed/HomeViewController.m]: in the DetailsViewController you can create a helper function set all of these values! 

`[detailVC setDetailName: nameToPass detailDetail:detailToPass detailLocation:cityToPass detailImage:imageToPass]`

File: Capstone/Models/College.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+@interface College : NSObject
+
+@property (strong, nonatomic) NSString *detailName;
+@property (strong, nonatomic) NSString *detailImage;
+@property (strong, nonatomic) NSString *detailLocation;
+@property (strong, nonatomic) NSString *detailDetail;
+
+
+- (instancetype)initWithDictionary:(NSArray *)dictionary;
+
```
Comment (skhezri) [inline Capstone/Models/College.h]: nit: extra line

File: Capstone/Models/College.h [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+NS_ASSUME_NONNULL_BEGIN
+
+@interface College : NSObject
+
+@property (strong, nonatomic) NSString *detailName;
+@property (strong, nonatomic) NSString *detailImage;
+@property (strong, nonatomic) NSString *detailLocation;
+@property (strong, nonatomic) NSString *detailDetail;
+
+
```
Comment (skhezri) [inline Capstone/Models/College.h]: nit: extra line

File: Capstone/Feed/HomeViewController.m [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    if (isFiltered) {
+        return self.filteredColleges.count;
+    }
+    return self.collegeDictionaryArrays.count;
+}
+
+- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
+    HomeCell *cell = [tableView dequeueReusableCellWithIdentifier:@"HomeCell"];
+    if (isFiltered) {
+        cell.homeCollegeName.text = self.filteredColleges[indexPath.row][@"name"];
```
Comment (skhezri) [inline Capstone/Feed/HomeViewController.m]: Do you plan to implement the College class here? definitely would help in terms of reducing redundant code!

**Review by skhezri — COMMENTED** [review]
Great work Abel!! Thanks for including a screenshot of the search bar, it looks awesome! I've added some comments regarding syntax and I think you have some opportunity to reduce redundant code in the next commit: detailsVC helper function, implementing college class with name, detail, city, and image. You've been making great progress on this project thus far!

File: Capstone/API/APIManager.m [hunk 11]
```diff
@@ -5,9 +5,19 @@
...
+
+
++ (instancetype)shared {
+    static APIManager *sharedManager = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedManager = [[self alloc] init];
+    });
+    return sharedManager;
+}
```
Comment (angelacaq) [inline Capstone/API/APIManager.m:22]: We discussed this a little bit, and I *think* I see the potential use case — but I am wondering what you think the purpose of this will be 🤔

File: Capstone/API/APIManager.h [hunk 12]
```diff
@@ -9,7 +9,11 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
-@interface APImanager : NSObject
+@interface APIManager : NSObject
+
++ (instancetype)shared;
+
+- (void)getHomeTimelineWithCompletion:(void(^)(NSArray *tweets, NSError *error))completion;
```
Comment (angelacaq) [inline Capstone/API/APIManager.h]: :0 what's this?

File: Capstone/Authentication/LoginViewController.m [hunk 13]
```diff
@@ -15,6 +15,8 @@
 @interface LoginViewController ()<RegistrationViewControllerDelegate>
 @property (weak, nonatomic) IBOutlet UITextField *usernameLogin;
 @property (weak, nonatomic) IBOutlet UITextField *passwordLogin;
+@property (weak, nonatomic) IBOutlet UIView *fbLogin;
```
Comment (angelacaq) [inline Capstone/Authentication/LoginViewController.m:18]: <3

File: Capstone/Authentication/LoginViewController.m [hunk 14]
```diff
@@ -38,33 +39,27 @@ - (IBAction)didTapLogin:(id)sender {
...
     NSString *username = self.usernameLogin.text;
     NSString *password = self.passwordLogin.text;
-    
     [PFUser logInWithUsernameInBackground:username password:password block:^(PFUser * user, NSError *  error) {
         if (error != nil) {
             NSLog(@"User log in failed: %@", error.localizedDescription);
         } else {
             NSLog(@"User logged in successfully");
-            
         }
```
Comment (angelacaq) [inline Capstone/Authentication/LoginViewController.m:48]: Perhaps add a TODO for error dialogs here

File: Capstone/Feed/HomeCell.h [hunk 15]
```diff
@@ -13,6 +13,8 @@ NS_ASSUME_NONNULL_BEGIN
 @property (weak, nonatomic) IBOutlet UIImageView *homeCellImage;
 @property (weak, nonatomic) IBOutlet UILabel *homeCollegeName;
 @property (weak, nonatomic) IBOutlet UILabel *homeCollegeDetails;
+@property (weak, nonatomic) IBOutlet UILabel *homeCollegeLocation;
+@property (weak, nonatomic) NSArray *cellDictionary;
```
Comment (angelacaq) [inline Capstone/Feed/HomeCell.h]: It'd be great if this array had a type so it's clear what objects go into this cell dictionary! I'm making my best guess — and the fact I have to guess is probably not good

```suggestion
@property (weak, nonatomic) NSArray<String *> *cellDictionary;
```

Also, what's the purpose of the dict? The name is a little vague

File: Capstone/Feed/HomeViewController.m [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+#import "HomeViewController.h"
+#import "HomeCell.h"
+#import "UIImageView+AFNetworking.h"
+#import "DetailsViewController.h"
+#import "APIManager.h"
+
+@interface HomeViewController ()<UITableViewDelegate, UITableViewDataSource, UISearchBarDelegate>
+
+@property(strong, nonatomic) NSArray<NSDictionary *> *colleges;
```
Comment (angelacaq) [inline Capstone/Feed/HomeViewController.m]: This is better! Why not use the College model we just created?

File: Capstone/Feed/HomeViewController.m [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+@property (strong, nonatomic) NSMutableArray *collegeDictionaryArrays;
+@property (weak, nonatomic) IBOutlet UISearchBar *searchBar;
+@property (strong, nonatomic) NSMutableArray *filteredColleges;
+
+@end
+
+bool isFiltered;
+
+@implementation HomeViewController
+
```
Comment (angelacaq) [inline Capstone/Feed/HomeViewController.m]: For private instance variables like this in Obj-C, you'll want to do something like this instead:

```suggestion
@implementation HomeViewController {
   bool _isFiltered;
}
```

Otherwise, you might have a global variable.

On second thought, for the non-outlets, you might want to also do the same:

```suggestion
@implementation HomeViewController {
   bool _isFiltered;
   NSArray<NSDictionary *> *_colleges;
   NSMutableArray<NSDictionary *> *_collegeDictionaryArrays;
   NSMutableArray<NSDictionary *> *_filteredColleges;
}
```

File: Capstone/Feed/HomeViewController.m [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+    [super viewDidLoad];
+    self.tableView.dataSource = self;
+    self.tableView.delegate = self;
+    self.searchBar.delegate = self;
+    isFiltered = false;
+    self.collegeDictionaryArrays = [[NSMutableArray alloc] init];
+    [self fetchData];
+}
+
+- (void)fetchData {
```
Comment (angelacaq) [inline Capstone/Feed/HomeViewController.m:36]: Bring back the TODO to move this to the APIManager?

or, significantly better, just move this to the APIManager 🥹

File: Capstone/Feed/HomeViewController.m [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+    [task resume];
+}
+
+-(void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
+    if (searchText.length == 0) {
+        isFiltered = false;
+    } else {
+        isFiltered = true;
+        self.filteredColleges = [[NSMutableArray alloc] init];
+        for (NSDictionary<NSString *, NSString *> *college in self.colleges) {
```
Comment (angelacaq) [inline Capstone/Feed/HomeViewController.m]: <3 yay types

File: Capstone/Feed/HomeViewController.m [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+        NSURL *url = [NSURL URLWithString:self.filteredColleges[indexPath.row][@"campusImage"]];
+        [cell.homeCellImage setImageWithURL:url];
+    } else {
+        cell.homeCollegeName.text = self.collegeDictionaryArrays[indexPath.row][@"name"];
+        cell.homeCollegeDetails.text = self.collegeDictionaryArrays[indexPath.row][@"shortDescription"];
+        cell.homeCollegeLocation.text = self.collegeDictionaryArrays[indexPath.row][@"city"];
+        NSURL *url = [NSURL URLWithString:self.collegeDictionaryArrays[indexPath.row][@"campusImage"]];
+        [cell.homeCellImage setImageWithURL:url];
+    }
+    return cell;
```
Comment (angelacaq) [inline Capstone/Feed/HomeViewController.m]: There's a lot of repetitive code that can be shortened to represent the same thing.

Often, when you notice that you're typing the same thing (`self.collegeDictionaryArrays[indexPath.row]`) multiple times, it might be better to make it a variable.

Additionally, both parts of the if/else statement do the same thing: the only difference is the dictionary itself. As such, we can do something like this:

```suggestion
    HomeCell *cell = [tableView dequeueReusableCellWithIdentifier:@"HomeCell"];
    NSDictionary<NSString *, NSString *> *college;
    if (isFiltered) {
       college = self.filteredColleges[indexPath.row];
    } else {
       college = self.collegeDictionaryArrays[indexPath.row];
    }
    cell.homeCollegeName.text = college[@"name"];
    cell.homeCollegeDetails.text = college[@"shortDescription"];
    cell.homeCollegeLocation.text = college[@"city"];
    NSURL *url = [NSURL URLWithString:college[@"campusImage"]];
    [cell.homeCellImage setImageWithURL:url];
    return cell;
```

File: Capstone/News/NewsViewController.m [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    self.tableView.dataSource = self;
+    self.tableView.delegate = self;
+    self.college1 = [[NSMutableArray alloc] init];
+    [self fetchData];
+}
+
+- (void)fetchData {
```
Comment (angelacaq) [inline Capstone/News/NewsViewController.m:29]: 😭😭😭😭😭😭😭😭😭😭😭😭😭😭😭😭😭

can we move the data fetching to the API manager? i think it was ok the first time (hence the earlier TODO comment), but now that it's in 3 places in the code, it's time to refactor

File: Capstone/Feed/HomeViewController.m [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+    NSIndexPath *myIndexPath = [self.tableView indexPathForCell:cell];
+    NSString *nameToPass = self.collegeDictionaryArrays[myIndexPath.row][@"name"];
+    NSString *detailToPass = self.collegeDictionaryArrays[myIndexPath.row][@"shortDescription"];
+    NSString *cityToPass = self.collegeDictionaryArrays[myIndexPath.row][@"city"];
+    NSString *imageToPass = self.collegeDictionaryArrays[myIndexPath.row][@"campusImage"];
+    DetailsViewController *detailVC = [segue destinationViewController];
+    detailVC.detailName = nameToPass;
+    detailVC.detailDetail = detailToPass;
+    detailVC.detailLocation = cityToPass;
+    detailVC.detailImage = imageToPass;
```
Comment (angelacaq) [inline Capstone/Feed/HomeViewController.m]: why not pass in the College model instead of each individual element?

File: Capstone/Feed/HomeViewController.m (hunk 20 above)
Comment (angelacaq) [reply Capstone/Feed/HomeViewController.m]: also, we could just pass in 
```
cell.college = college;
```
if the arrays contained College objects, and have the HomeCell deal with the image setting 🤔

File: Capstone/Feed/HomeViewController.m [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+#import "HomeCell.h"
+#import "UIImageView+AFNetworking.h"
+#import "DetailsViewController.h"
+#import "APIManager.h"
+
+@interface HomeViewController ()<UITableViewDelegate, UITableViewDataSource, UISearchBarDelegate>
+
+@property(strong, nonatomic) NSArray<NSDictionary *> *colleges;
+@property (weak, nonatomic) IBOutlet UITableView *tableView;
+@property (strong, nonatomic) NSMutableArray *collegeDictionaryArrays;
```
Comment (angelacaq) [inline Capstone/Feed/HomeViewController.m]: what makes this different from the `colleges` array?

**Review by angelacaq — COMMENTED** [review]
Woohoo, love to see a test plan!

Since this is just the second PR after a lot of issues yesterday, I'm not going to block changes here; however, starting next week, this PR would be something I'd request changes for 😔

As we get more comfortable with PRs, keeping them small per feature will be super helpful for reviewers and for you when applying reviewer comments -- because I unfortunately have a lot of comments.


## data model

On this PR, the major architecture issue I see is the creation of a `College` object which should be acting as the data model we're passing around. Yet, we're also using dictionaries, which represent the same thing as `College`, except we have to manually input the name, short description, etc. every time we want to pass around college info from one place or another.

Basically, instead of dictionaries, why can't we use the `College` model?


## refactoring

Now that the fetchData method, with pretty much the same code, is used in multiple areas, it's a good time to migrate this over to APIManager. It's better to refactor while ahead instead of on cleanup, since refactoring later can often cause things to break 😖
